File: core/src/es/danirod/rectball/actors/StatsTable.java
package es.danirod.rectball.actors;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle;
import com.badlogic.gdx.scenes.scene2d.utils.Drawable;
import com.badlogic.gdx.utils.Align;

import java.util.*;

import es.danirod.rectball.RectballGame;
import es.danirod.rectball.statistics.StatisticSet;
import es.danirod.rectball.statistics.Statistics;

public class StatsTable extends Table {

    private RectballGame game;

    private Statistics stats;

    private LabelStyle title;

    private LabelStyle data;

    public StatsTable(RectballGame game, LabelStyle title, LabelStyle data) {
        this.game = game;
        this.stats = game.statistics;
        this.title = title;
        this.data = data;

        defaults().padBottom(30);
        add(addBestScores()).fillX().expandX().row();
        add(addTotalData()).fillX().expandX().row();
        add(addColorData()).fillX().expandX().row();
        add(addSizesData()).fillX().expandX().row();
    }

    private Table addBestScores() {
        Table best = new Table();
        best.add(new Label(game.getLocale().get("statistics.best"), this.title)).colspan(2).row();

        boolean printedSomething = false;
        // Add best score.
        if (game.scores.getHighestScore() != 0) {
            String bestScore = Long.toString(game.scores.getHighestScore());
            best.add(new Label(game.getLocale().get("statistics.best.score"), data)).align(Align.left).fillX();
            best.add(new Label(bestScore, data)).align(Align.right).expandX().row();
            printedSomething = true;
        }

        // Add best time.
        if (game.scores.getHighestTime() != 0) {
            String bestTime = secondsToTime(game.scores.getHighestTime());
            best.add(new Label(game.getLocale().get("statistics.best.time"), data)).align(Align.left).fillX();
            best.add(new Label(bestTime, data)).align(Align.right).expandX().row();
            printedSomething = true;
        }

        // If everything is 0, don't print anything.
        if (!printedSomething) {
            Label noData = new Label(game.getLocale().get("statistics.noData"), game.getSkin());
            noData.setAlignment(Align.center);
            best.add(noData).colspan(2).fillX().expandX().padTop(10).padBottom(10).row();
        }

        return best;
    }

    private Table addTotalData() {
        Table total = new Table();
        total.add(new Label(game.getLocale().get("statistics.total"), this.title)).colspan(2).row();

        StatisticSet set = game.statistics.getTotalData();
        if (set.getStats().isEmpty()) {
            Label noData = new Label(game.getLocale().get("statistics.noData"), game.getSkin());
            noData.setAlignment(Align.center);
            total.add(noData).colspan(2).fillX().expandX().padTop(10).padBottom(10).row();
            return total;
        }

        for (Map.Entry<String, Integer> stat : set.getStats().entrySet()) {
            String statName = game.getLocale().get("statistics.total." + stat.getKey());
            String statValue = Integer.toString(stat.getValue());

            if (stat.getKey().equals("time")) {
                statValue = secondsToTime(stat.getValue());
            }

            total.add(new Label(statName, data)).align(Align.left).fillX();
            total.add(new Label(statValue, data)).align(Align.right).expandX().row();
        }

        return total;
    }

    private Table addColorData() {
        Table color = new Table();
        color.add(new Label(game.getLocale().get("statistics.color"), this.title)).colspan(4).row();

        if (game.statistics.getColorData().getStats().isEmpty()) {
            Label noData = new Label(game.getLocale().get("statistics.noData"), game.getSkin());
            noData.setAlignment(Align.center);
            color.add(noData).colspan(4).fillX().expandX().padTop(10).padBottom(10).row();
            return color;
        }

        // Put the data in a TreeMap. TreeMaps keep order. Use reverse to show biggers on left.
        Map<Integer, String> colorScore = new TreeMap<>(Collections.reverseOrder());
        colorScore.put(game.statistics.getColorData().getValue("red"), "red");
        colorScore.put(game.statistics.getColorData().getValue("green"), "green");
        colorScore.put(game.statistics.getColorData().getValue("blue"), "blue");
        colorScore.put(game.statistics.getColorData().getValue("yellow"), "yellow");

        color.defaults().expandX().fillX().align(Align.center).size(60).padTop(5);
        for (Map.Entry<Integer, String> entry : colorScore.entrySet()) {
            color.add(new Image(game.getBallAtlas().findRegion("ball_" + entry.getValue())));
        }
        color.row();
        for (Map.Entry<Integer, String> entry : colorScore.entrySet()) {
            Label label = new Label(entry.getKey().toString(), data);
            label.setAlignment(Align.center);
            color.add(label);
        }
        color.row();

        return color;
    }

    private Table addSizesData() {
        Table sizes = new Table();
        sizes.add(new Label(game.getLocale().get("statistics.sizes"), this.title)).colspan(3).row();

        Map<String, Integer> unsortedScores = game.statistics.getSizesData().getStats();
        if (unsortedScores.isEmpty()) {
            Label noData = new Label(game.getLocale().get("statistics.noData"), game.getSkin());
            noData.setAlignment(Align.center);
            sizes.add(noData).colspan(3).fillX().expandX().padTop(10).padBottom(10).row();
            return sizes;
        }

        Map<Integer, String> sizesScore = new TreeMap<>(Collections.reverseOrder());
        for (Map.Entry<String, Integer> score : unsortedScores.entrySet()) {
            sizesScore.put(score.getValue(), score.getKey());
        }

        int highestValue = -1;
        Drawable bar = game.getSkin().newDrawable("pixel", Color.WHITE);
        for (Map.Entry<Integer, String> score : sizesScore.entrySet()) {
            // Put the highest value. This should be done using the first item.
            if (highestValue == -1) {
                highestValue = score.getKey();
            }

            sizes.add(new Label(score.getValue(), data)).align(Align.left).fillX();
            sizes.add(new Label(score.getKey().toString(), data)).align(Align.right).padRight(10).expandX();

            // Add a bar to the right
            float percentage = (float) score.getKey() / highestValue;
            sizes.add(new Image(bar)).width(240 * percentage).padLeft(240 * (1 - percentage)).padBottom(5).fill().row();
        }

        return sizes;
    }

    private static String secondsToTime(int seconds) {
        if (seconds < 60) {
            return Integer.toString(seconds);
        } else {
            String min = Integer.toString(seconds / 60);
            String sec = Integer.toString(seconds % 60);
            if (sec.length() == 1) {
                sec = "0" + sec;
            }
            return min + ":" + sec;
        }
    }
}

File: core/src/main/java/es/danirod/rectball/scene2d/ui/StatsTable.kt
/* This file is part of Rectball
 * Copyright (C) 2015-2024  Dani Rodríguez
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package es.danirod.rectball.scene2d.ui

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.utils.Align
import es.danirod.rectball.RectballGame
import kotlin.time.Duration.Companion.seconds

class StatsTable(private val game: RectballGame, private val title: LabelStyle, private val data: LabelStyle) : Table() {

    private val context = game.context

    init {
        defaults().padBottom(30f)
        add(addBestScores()).fillX().expandX().row()
        add(addTotalData()).fillX().expandX().row()
        add(addColorData()).fillX().expandX().row()
        add(addSizesData()).fillX().expandX().row()
    }

    private fun addBestScores(): Table {
        val best = Table()

        best.add(Label(game.locale["statistics.best_data"], this.title)).colspan(2).row()

        if (game.statistics.highScore == 0L && game.statistics.highTime == 0L) {
            val noData = Label(game.locale["statistics.no_data"], game.appSkin)
            noData.setAlignment(Align.center)
            best.add(noData).colspan(2).fillX().expandX().padTop(10f).padBottom(10f).row()
            return best
        }

        val highScore = DecimalValue(game.statistics.highScore)
        if (highScore.hasValue()) append(best, game.locale["statistics.best_score"], highScore.toString())

        val highTime = TimeValue(game.statistics.highTime)
        if (highTime.hasValue()) append(best, game.locale["statistics.best_time"], highTime.toString())

        return best
    }

    private fun addTotalData(): Table {
        val total = Table()
        total.add(Label(game.locale["statistics.total_data"], title)).colspan(2).row()
        val data = totalStatistics
        if (data.isEmpty()) {
            val noData = Label(game.locale["statistics.no_data"], game.appSkin)
            noData.setAlignment(Align.center)
            total.add(noData).colspan(2).fillX().expandX().padTop(10f).padBottom(10f).row()
        } else {
            data.forEach { (key, value) -> append(total, key, value.toString()) }
        }
        return total
    }

    private fun addColorData(): Table {
        val color = Table()

        val stats = sortStatsMap(game.statistics.colorStatistics)

        if (stats.isEmpty()) {
            color.add(Label(game.locale["statistics.color_data"], title)).row()
            val noData = Label(game.locale["statistics.no_data"], game.appSkin)
            noData.setAlignment(Align.center)
            color.add(noData).fillX().expandX().padTop(10f).padBottom(10f).row()
        } else {
            color.add(Label(game.locale["statistics.color_data"], title)).colspan(stats.size).row()
            color.defaults().expandX().fillX().align(Align.center).size(60f).padTop(5f)

            for ((key, _) in stats) { color.add(Image(game.ballAtlas.findRegion("ball_$key"))) }
            color.row()
            for ((_, value) in stats) {
                val label = Label(value.toString(), data)
                label.setAlignment(Align.center)
                color.add(label)
            }
            color.row()
        }

        return color
    }

    private fun addSizesData(): Table {
        val sizes = Table()
        val stats = sortStatsMap(game.statistics.sizeStatistics)

        sizes.add(Label(game.locale["statistics.size_data"], this.title)).colspan(3).row()

        /* No data. */
        if (stats.isEmpty()) {
            val noData = Label(game.locale["statistics.no_data"], game.appSkin)
            noData.setAlignment(Align.center)
            sizes.add(noData).colspan(3).fillX().expandX().padTop(10f).padBottom(10f).row()
            return sizes
        }

        val bar = game.appSkin.newDrawable("pixel", Color.WHITE)

        // Highest value is used to calculate the relative percentage of each row.
        val highestValue = stats.first().second

        for ((key, value) in stats) {
            val percentage = value.toFloat() / highestValue
            sizes.add(Label(key, data)).align(Align.left).fillX()
            sizes.add(Label(value.toString(), data)).align(Align.right).padRight(10f).expandX()
            sizes.add(Image(bar)).width(240 * percentage).padLeft(240 * (1 - percentage)).padBottom(5f).fill().row()
        }

        return sizes
    }

    /** A map that pairs an statistic label into the value for that statistic label. */
    private val totalStatistics: Map<String, Value>
        get() = mapOf(
            game.locale["statistics.total_score"] to DecimalValue(game.statistics.totalScore),
            game.locale["statistics.total_combinations"] to DecimalValue(game.statistics.totalCombinations),
            game.locale["statistics.total_gems"] to DecimalValue(game.statistics.totalGems),
            game.locale["statistics.total_games"] to DecimalValue(game.statistics.totalGames),
            game.locale["statistics.total_time"] to TimeValue(game.statistics.totalTime),
            game.locale["statistics.total_perfect"] to DecimalValue(game.statistics.totalPerfects),
            game.locale["statistics.total_hints"] to DecimalValue(game.statistics.totalHints),
        ).filterValues { it.hasValue() }

    /** Sorts a [statistics] map by score, highest scores come first. */
    private fun sortStatsMap(statistics: Map<String, Long>) =
            statistics.toList().sortedBy { it.second }.reversed()

    /** Appends a statistic composed by the label [name] and the value [value] to a [table]. */
    private fun append(table: Table, name: String, value: String) {
        table.add(Label(name, data)).align(Align.left).fillX()
        table.add(Label(value, data)).align(Align.right).expandX().row()
    }

    private open class Value(val value: Long) {
        fun hasValue() = value > 0
    }

    private class DecimalValue(value: Long): Value(value) {
        override fun toString() = value.toString()
    }

    private class TimeValue(value: Long): Value(value) {
        override fun toString() = value.seconds.toComponents { hh, mm, ss, _ ->
            when {
                hh > 0 -> String.format("%d:%02d:%02d h", hh, mm, ss)
                mm > 0 -> String.format("%d:%02d m", mm, ss)
                else -> String.format("%d s", ss)
            }
        }
    }
}
